src/libteam/patch/0015-Extend-LACP-fallback-support-to-multiple-ports.patch

```diff
@@ -0,0 +1,253 @@
+-	    is_lacp_fallback_eligible(lacp_port)))
++	    lacp_port->state == PORT_STATE_FALLBACK))
+ 		return true;
+ 	return false;
+ }
+@@ -582,7 +570,7 @@ static int lacp_port_should_be_disabled(struct lacp_port *lacp_port)
+ 	if (!lacp_port_selected(lacp_port) ||
+ 	    lacp_port->agg_lead != lacp->selected_agg_lead ||
+ 	    (!(lacp_port->partner.state & INFO_STATE_SYNCHRONIZATION) &&
+-	    !is_lacp_fallback_eligible(lacp_port)))
++	    lacp_port->state != PORT_STATE_FALLBACK))
+ 		return true;
+ 	return false;
+ }
+@@ -1249,7 +1237,7 @@ static void lacp_port_actor_update(struct lacp_port *lacp_port)
+ 	}
+ 	if (lacp_port->state == PORT_STATE_EXPIRED)
+ 		state |= INFO_STATE_EXPIRED;
+-	if (lacp_port->state == PORT_STATE_DEFAULTED)
++	if (lacp_port->state == PORT_STATE_DEFAULTED || lacp_port->state == PORT_STATE_FALLBACK)
+ 		state |= INFO_STATE_DEFAULTED;
+ 	if (teamd_port_count(lacp_port->ctx) > 0)
+ 		state |= INFO_STATE_AGGREGATION;
+@@ -1258,6 +1246,101 @@ static void lacp_port_actor_update(struct lacp_port *lacp_port)
+ 
+ static int lacpdu_send(struct lacp_port *lacp_port);
+ 
++static int lacp_port_set_state(struct lacp_port *lacp_port,
++			       enum lacp_port_state new_state);
++
++static bool lacp_port_better_for_fallback(struct lacp_port *lacp_port1, struct lacp_port *lacp_port2)
++{
++	int system_diff;
++	struct lacpdu_info *actor1;
++	struct lacpdu_info *actor2;
++
++	if (!lacp_port2)
++		return true;
++
++	actor1 = &lacp_port1->actor;
++	actor2 = &lacp_port2->actor;
++
++	system_diff = memcmp(actor1->system, actor2->system, ETH_ALEN);
++	/* If system MACs differ, the port with the lower system wins. */
++	if (system_diff != 0)
++		return system_diff < 0;
++
++	/* If port priorities differ, the port with the higher priority wins. */
++	if (actor2->port_priority != actor1->port_priority)
++		return (ntohs(actor1->port_priority) > ntohs(actor2->port_priority));
++
++	/* Otherwise, the port with the lower number wins. */
++	return ntohs(actor1->port) < ntohs(actor2->port);
++}
++
++static int lacp_ports_update_fallback_state(struct lacp *lacp)
++{
++	struct lacp_port *current_fallback = NULL;
++	struct lacp_port *new_fallback = NULL;
++	struct teamd_port *tdport;
++	bool have_selected_port = false;
++
++	/* If fallback is disabled, no need to do anything. */
++	if (!lacp->cfg.fallback) {
++		return 0;
++	}
++
++	/* Check all ports.
++	 * If any port is receiving LACPDUs, we disable fallback altogether.
++	 * Otherwise, the non-disabled port with the highest priority is moved
++	 * to fallback mode. */
++	teamd_for_each_tdport(tdport, lacp->ctx) {
++		struct lacp_port *lacp_port = lacp_port_get(lacp, tdport);
++
++		if (lacp_port->state == PORT_STATE_DISABLED) {
++			/* Ignore disabled ports. */
++			continue;
++		}
++		if (lacp_port->state == PORT_STATE_CURRENT || lacp_port->state == PORT_STATE_EXPIRED) {
++			/* If at least one port is currently not defaulted, don't do any fallback. */
++			have_selected_port = true;
```

Comment:
Could you rename this to something like `have_active_port` or similar? `have_selected_port` sounds like a port was selected in this loop.

Reply:
I'm not certain that `selected` is perfectly accurate here, so I've renamed this to `do_fallback`, which is a better description of the purpose anyways - it's supposed to track whether we want to use fallback logic or not.